napari/util/keybindings.py

```diff
@@ -0,0 +1,372 @@
+    Due to the native implementation of the key system, Shift pressed in certain
+    key combinations may yield inconsistent or unexpected results.
+    Therefore, it is not recommended to use Shift with non-letter keys.
+    On OSX, Control is swapped with Meta such that pressing Command reads as
+    Control.
+
+    Parameters
+    ----------
+    comb : str
+        Key combination.
+
+    Returns
+    -------
+    normalized_comb : str
+        Normalized key combination.
+    """
+    key, modifiers = parse_comb(comb)
+
+    if len(key) != 1 and key not in SPECIAL_KEYS:
+        raise TypeError(f'invalid key {key}')
+
+    for modifier in modifiers:
+        if modifier not in MODIFIER_KEYS:
+            raise TypeError(f'invalid modifier key {modifier}')
+
+    return components_to_comb(key, modifiers)
+
+
+UNDEFINED = object()
+
+
+def bind_key(keymap, key, func=UNDEFINED, *, overwrite=False):
+    """Bind a key combination to a keymap.
+
+    Parameters
+    ----------
+    keymap : dict of str: callable
+        Keymap to modify.
+    key : str
+        Key combination.
+    func : callable or None
+        Callable to bind to the key combination.
+        If ``None`` is passed, unbind instead.
+    overwrite : bool, keyword-only, optional
+        Whether to overwrite the key combination if it already exists.
+
+    Returns
+    -------
+    unbound : callable or None
+        Callable unbound by this operation, if any.
+
+    Notes
+    -----
+    Keys are represented in the form ``[modifier-]key``, e.g. ``a``,
+    ``Control-c``, or ``Control-Alt-Delete``.
+    Valid modifiers are Control, Alt, Shift, and Meta.
+
+    Letters will always be read as upper-case.
+    Due to the native implementation of the key system, Shift pressed in certain
+    key combinations may yield inconsistent or unexpected results.
+    Therefore, it is not recommended to use Shift with non-letter keys.
+    On OSX, Control is swapped with Meta such that pressing Command reads as
+    Control.
+
+    Special keys include Shift, Control, Alt, Meta, Up, Down, Left, Right,
+    PageUp, PageDown, Insert, Delete, Home, End, Escape, Backspace, F1,
+    F2, F3, F4, F5, F6, F7, F8, F9, F10, F11, F12, Space, Enter, and Tab
+
+    Functions take in only one argument: the parent that the function
+    was bound to. This is the viewer or layer.
+
+    By default, all functions are assumed to work on key presses only,
+    but can be denoted to work on release too by separating the function
+    into two statements with the yield keyword::
+
+        @viewer.bind_key('h')
+        def hello_world(viewer):
+            # on key press
+            viewer.status = 'hello world!'
+
+            yield
+
+            # on key release
+            viewer.status = 'goodbye world :('
+
+    """
+    if func is UNDEFINED:
+
+        def inner(func):
+            bind_key(keymap, key, func, overwrite=overwrite)
+            return func
+
+        return inner
```

Comment:
What's happening here? Can you explain it in a comment?

Reply:
it's similar to using `functools.partial` except it returns a function, which allows this to be used as a decorator with arguments and also as a regular function

Reply:
I'm confused. functools.partial also returns a function?

Reply:
Ah yes, I should've used better language ;P

So I'm sure you're familiar with decorators, but to explain this aspect, it's time to go on a deep-dive *queue theme music* 😎 (let's ignore that I'm just generally bad at concisely explaining things)

So I'm sure you're familiar with simple decorators which take in a single argument, the function, do something with it, and either return the same function or a modified version. The type we're using is registration so I won't go into modifying them but you can find examples at projects like [`wrapt`](https://wrapt.readthedocs.io).

For example, here is a basic registration decorator
```python
special_func_registry = []

def special_func(func):
    special_func_registry.append(func)
    return func

@special_func
def foo(): ...
```

However, say you wanted to allow users to specify a name for the function:
```python
special_func_registry = {}

def register_special_func(name):
    def inner(func):  # this is the actual decorator
        special_func_registry[name] = func
        return func
    return inner

@register_special_func('bar')
def foo(): ...
```

What happens here is that `register_special_func('bar')` resolves to `inner` which is then applied as a decorator.

Now, say that you wanted to also use this as a regular function instead of as a decorator. You'd have to do something like `register_special_func('bar')(foo)`, which is pretty nasty.

So we revise it again:
```python
def register_special_func(name, func=None):
    if func is not None:  # used as a function
        special_func_registry[name] = func
    else:                 # used as a decorator
        def inner(func):
            register_special_func(name, func)
            return func
        return inner
```

In cases where the functional version of the decorator does not return anything, you can use `functools.partial` and have it return the decorated function:
```python
def register_special_func(name, func=None):
    if func is None:
        return functools.partial(register_special_func, name)
    special_func_registry[name] = func
    return func
```

However, if you were to try this while returning anything other than the decorated function, you would be met with disaster:
```python
def register_special_func(name, func=None):
    if func is None:
        return functools.partial(register_special_func, name)
    special_func_registry[name] = func
    return name

@register_special_func('bar')
def foo(): ...



>>> foo()
TypeError: 'str' object is not callable

>>> foo
'bar'
```

This is why I implemented it the way I did: it allows you to both use `bind_key` as a function and a decorator. As a function, `bind_key` will return any function unbound by the operation performed (whether it was by explicitly unbinding the function or overwriting it) while also not terrifyingly deleting all your functions by accident.

Reply:
ooooooh!

I *think* what you want is `toolz.curry`. SOOOO much simpler:

```python
import toolz as tz

@tz.curry
def register_special_func(name, func):
    special_func_registry[name] = func
    return func
```

Reply:
`toolz.curry` has the same problem as using `functools.partial` as explained in the last section above